### ISSUE TITLE
Indexing a Type Alias of an Enum yields false error

### ISSUE THREAD

**Issue by brianjyoung** [python/mypy#7568]
### Minimum [not-]working product:
Ah, good old Enums:
```
from enum import Enum
class EnumClass(Enum):
    one = 1

EnumAlias = EnumClass
EnumAlias['one']
```
```
test.py:6: error: Type application targets a non-generic function or class
test.py:6: error: Name 'one' is not defined
```
I'm _fairly certain_ this is not an intentional design choice. Just to be sure, according to documentation:
> A type alias does not create a new type. It’s just a shorthand notation for another type – it’s equivalent to the target type except for generic aliases.

Now, Enums actually _can't_ be generic, so I certainly expect the alias to be equivalent. So I think this code should pass (regardless of its usefulness... who wants to alias an Enum anyways?).

### Possible Causes
I'm no expert, but I think this is happening because the semantic analyzer deems any `IndexExpr` a `TypeApplication` if it has a `TypeAlias` as a base. Once it's marked as analyzed, it doesn't reach the stage of the expression checker where the Enum type would be recognized and the indexing processed as valid. Whatever the root of the problem, I think there's only one spot where the expression checker actually looks at `is_enum` to handle Enum indexing, so the alias isn't making it there one way or another.

**Comment by ilevkivskyi** [comment]
I think we can disambiguate this with a bunch special-casing, but it looks like a low priority one. Also using `EnumAlias.one` works.

**Comment by colincadams** [comment]
Is there a reasonable way to work around this?

**Comment by hauntsaninja** [comment]
`EnumAlias.one` as Ivan mentioned seems like a reasonable workaround.

**Comment by hauntsaninja** [comment]
```
from typing import Type
from enum import Enum
class EnumClass(Enum):
    one = 1

EnumAlias: Type[EnumClass] = EnumClass
reveal_type(EnumClass['one'])
reveal_type(EnumAlias['one'])
reveal_type(EnumClass.one)
reveal_type(EnumAlias.one)
```
This seemed to do what I'd expect, not sure if there are any sharp edges still lurking.

**Comment by incidentist** [comment]
I looked at the [diff](https://github.com/y0-causal-inference/y0/commit/2c837b73a090cf0058f9de25ba60470f77b5950c#diff-0b7457aea011612c11d65101a25eee3d914a8a0739ef9e82aa489e700660bd27) in the above reference, and it seems like they have a good workaround that involves declaring the type of the alias. Here is a version of the repro code that mypy finds no issue with:

```
from enum import Enum
from typing import Type


class EnumClass(Enum):
    one = 1


EnumAlias: Type[EnumClass] = EnumClass # <-- fixed line
EnumAlias["one"]
```

**Comment by incidentist** [comment]
Update: the above workaround validates the alias for member access and subscripting, but breaks it as a type annotation. In our case, we do a bunch of hacky stuff to get the alias working right in runtime, and then we also add this to get it validating in mypy:
```
if typing.TYPE_CHECKING:
    # Importing the class under the alias name (rather than assigning
    # `EnumAlias = EnumClass`) makes mypy treat `EnumAlias` as a full re-binding
    # of the class -- so it is BOTH a value (subscript/call/iteration) AND a
    # valid type annotation. A plain assignment alias collapses to one or the
    # other (see https://github.com/python/mypy/issues/7568).
    # pylint: disable=import-self
    from test import EnumClass as EnumAlias
```
where `test.py` is the current filename. This works for us but a) is a hack and b) is very dependent on how your enums are set up and used. But if you are running into this issue, it's worth it to play around with that kind of self-import under the `typing.TYPE_CHECKING` flag to see if you can find a workaround.
